### PR TITLE
Implement ZSink#collectAllWhileWith (#4886)

### PR DESCRIPTION
**Pull request by jupposessho**

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/experimental/ZSink.scala [hunk 1]
```diff
@@ -77,7 +77,35 @@ class ZSink[-R, -InErr, -In, +OutErr, +L, +Z](val channel: ZChannel[R, InErr, Ch
...
+          def loop(currentResult: S): ZChannel[R, InErr, Chunk[In], Any, OutErr, Chunk[L], S] =
+            channel.doneCollect
+              .foldChannel(
+                ZChannel.fail(_),
+                { case (leftovers, ze) =>
+                  if (p(ze)) {
+                    for {
+                      _            <- ZChannel.fromZIO(leftoversRef.set(leftovers.flatten.asInstanceOf[Chunk[In]]))
+                      upstreamDone <- ZChannel.fromZIO(upstreamDoneRef.get)
+                      result       <- if (upstreamDone) ZChannel.end(f(currentResult, ze)) else loop(f(currentResult, ze))
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/experimental/ZSink.scala:100]: Write the leftovers before terminating:

```suggestion
                      result       <- if (upstreamDone) ZChannel.write(leftovers.flatten) *> ZChannel.end(f(currentResult, ze)) 
                                           else loop(f(currentResult, ze))
```

**Review by iravid — COMMENTED** [review]
Looks great!

File: streams/shared/src/main/scala/zio/stream/experimental/ZSink.scala (hunk 1 above)
Comment (jupposessho) [reply streams/shared/src/main/scala/zio/stream/experimental/ZSink.scala]: fixed, thanks!

**Review by iravid — APPROVED** [review]
Great work @jupposessho

**Pull request merged**